### PR TITLE
IRSA module refactor (No Breaking changes)

### LINKED ISSUE #697: [FEATURE] - Simplify usage of kubernetes-addons and IRSA modules

### PR DESCRIPTION
**Pull request by vara-bonthu**
### What does this PR do?

This PR is raised to simplify the IRSA module variables. No breaking changes are introduced by this PR to EKS blueprints. 
However, users who are calling IRSA module directly needs a simple change to your code. Check the AFTER code snippet. FAQ.md also provides these details

## Before 

This is the current usage of consuming IRSA module directly.

```hcl

module "irsa" {
  source = "github.com/aws-ia/terraform-aws-eks-blueprints//modules/irsa"
  kubernetes_namespace       = "dev"
  kubernetes_service_account = "spark"
  irsa_iam_policies          = [aws_iam_policy.spark.arn]
  addon_context = {
    eks_cluster_id                = var.eks_cluster_id
    eks_oidc_provider_arn         = "arn:${data.aws_partition.current.partition}:iam::${data.aws_caller_identity.current.account_id}:oidc-provider/${local.eks_oidc_issuer_url}"
    eks_oidc_issuer_url           = local.eks_oidc_issuer_url
    irsa_iam_role_path            = "/"
    irsa_iam_permissions_boundary = ""
    aws_caller_identity_account_id = ""
    aws_caller_identity_arn = ""
    aws_eks_cluster_endpoint = ""
    aws_partition_id = ""
    aws_region_name = ""
    tags = {}
  }
}
```

## After

```hcl
module "irsa" {
  source = "github.com/aws-ia/terraform-aws-eks-blueprints//modules/irsa"
  kubernetes_namespace       = "dev"
  kubernetes_service_account = "spark"
  irsa_iam_policies          = [aws_iam_policy.spark.arn]
  eks_cluster_id             = module.eks_blueprints.eks_cluster_id
  eks_oidc_provider_arn      = module.eks_blueprints.eks_oidc_provider_arn
}
```

Related Issues
closes #697 
closes #814

### Motivation

- To simplify user experience to consume IRSA module directly.
- There are number of instances where users want to create IRSA for specific service account

### More

- [x] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)
- [ ] Yes, I have added a new example under [examples](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [x] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators
- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: modules/irsa/main.tf [hunk 1]
```diff
@@ -20,31 +20,31 @@ resource "kubernetes_service_account_v1" "irsa" {
...
         "Principal" : {
-          "Federated" : "${var.addon_context.eks_oidc_provider_arn}"
+          "Federated" : "${var.eks_oidc_provider_arn}"
         },
         "Action" : "sts:AssumeRoleWithWebIdentity",
         "Condition" : {
           "StringLike" : {
-            "${var.addon_context.eks_oidc_issuer_url}:sub" : "system:serviceaccount:${var.kubernetes_namespace}:${var.kubernetes_service_account}",
-            "${var.addon_context.eks_oidc_issuer_url}:aud" : "sts.amazonaws.com"
+            "${var.eks_oidc_issuer_url}:sub" : "system:serviceaccount:${var.kubernetes_namespace}:${var.kubernetes_service_account}",
```
Comment (bryantbiggs) [inline modules/irsa/main.tf:36]: optional: you could do something similar to [this](https://github.com/terraform-aws-modules/terraform-aws-iam/blob/master/modules/iam-role-for-service-accounts-eks/main.tf#L18) where you can remove the `var.eks_oidc_issuer_url` variable and use the `var.eks_oidc_provider_arn` like:

```suggestion
            "${replace(var.eks_oidc_provider_arn, "/^(.*provider/)/", "")}:sub" : "system:serviceaccount:${var.kubernetes_namespace}:${var.kubernetes_service_account}",
```

**Review by bryantbiggs — APPROVED** [review]
Looks great!

File: modules/irsa/main.tf (hunk 1 above)
Comment (vara-bonthu) [reply modules/irsa/main.tf:36]: This is much better 👍🏼  Let me test this

**Comment by bryantbiggs** [comment]
`plan-examples / Plan examples (examples/external-secrets-kubernetes-addon) (pull_request_target) Failing after 8s — Plan examples (examples/external-secrets-kubernetes-addon) ` will fail due to name change, ok to ignore

File: modules/irsa/main.tf [hunk 2]
```diff
@@ -1,3 +1,7 @@
+locals {
+  eks_oidc_issuer_url = "${replace(var.eks_oidc_provider_arn, "/^(.*provider/)/", "")}:sub"
```
Comment (Zvikan) [inline modules/irsa/main.tf:2]: @vara-bonthu should this have `:sub` ?

Comment (srhoton) [reply modules/irsa/main.tf:2]: This is breaking things, I believe. Look at https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/837

**Pull request merged**